opm/ml/ml_model.hpp

```diff
@@ -0,0 +1,437 @@
+    OPM_ERROR_IF(!(j < dims_[1] && j >= 0), fmt::format(" Invalid j:  "
+                                                        "{}"
+                                                        "  max: "
+                                                        "{}",
+                                                        j, dims_[1]));
+    OPM_ERROR_IF(!(k < dims_[2] && k >= 0), fmt::format(" Invalid k:  "
+                                                        "{}"
+                                                        "  max: "
+                                                        "{}",
+                                                        k, dims_[2]));
+    OPM_ERROR_IF(!(l < dims_[3] && l >= 0), fmt::format(" Invalid l:  "
+                                                        "{}"
+                                                        "  max: "
+                                                        "{}",
+                                                        l, dims_[3]));
+
+    return data_[dims_[3] * (dims_[2] * (dims_[1] * i + j) + k) + l];
+  }
+
+  const T& operator()(int i, int j, int k, int l) const {
+    OPM_ERROR_IF(dims_.size() != 4, "Invalid indexing for tensor");
+    OPM_ERROR_IF(!(i < dims_[0] && i >= 0), fmt::format(" Invalid i:  "
+                                                        "{}"
+                                                        "  max: "
+                                                        "{}",
+                                                        i, dims_[0]));
+    OPM_ERROR_IF(!(j < dims_[1] && j >= 0), fmt::format(" Invalid j:  "
+                                                        "{}"
+                                                        "  max: "
+                                                        "{}",
+                                                        j, dims_[1]));
+    OPM_ERROR_IF(!(k < dims_[2] && k >= 0), fmt::format(" Invalid k:  "
+                                                        "{}"
+                                                        "  max: "
+                                                        "{}",
+                                                        k, dims_[2]));
+    OPM_ERROR_IF(!(l < dims_[3] && l >= 0), fmt::format(" Invalid l:  "
+                                                        "{}"
+                                                        "  max: "
+                                                        "{}",
+                                                        l, dims_[3]));
+
+    return data_[dims_[3] * (dims_[2] * (dims_[1] * i + j) + k) + l];
+  }
+  void fill(const T& value) { std::fill(data_.begin(), data_.end(), value); }
+
+  // Tensor addition
+  Tensor operator+(const Tensor& other) {
+    OPM_ERROR_IF(dims_.size() != other.dims_.size(),
+                 "Cannot add tensors with different dimensions");
+    Tensor result;
+    result.dims_ = dims_;
+    result.data_.reserve(data_.size());
```

Comment:
This should be a resize(), and then not use a back_inserter below, as discussed earlier! Same with the multiply() further down.